File: tools/synthesia-extract/synthesia_extract/schema.py
"""M6-A ``ExtractedChart`` schema, mirrored in Python.

The JSON these emit is the wire format read by ``crates/import`` (see
``crates/import/src/schema.rs``).  Field names, the ``Hand`` string variants
(``"Left"`` / ``"Right"`` / ``"Unknown"``), and the ``skip_serializing_if =
Option::is_none`` behaviour for ``velocity`` / ``confidence`` / ``title`` /
``fps`` / ``scroll_px_per_s`` are all reproduced so the Rust ``serde``
deserializer accepts the output verbatim.
"""

from __future__ import annotations

import enum
import json
from dataclasses import dataclass, field
from typing import Any, Optional

# Bump when the extraction algorithm changes in a way reviewers should notice.
EXTRACTOR_VERSION = "synthesia-extract 0.2.0"


class Hand(enum.Enum):
    """Which hand played a note. Serializes to the Rust enum's variant name."""

    LEFT = "Left"
    RIGHT = "Right"
    UNKNOWN = "Unknown"

    def __str__(self) -> str:  # pragma: no cover - convenience only
        return self.value


@dataclass
class ExtractedNote:
    """One detected note. Mirrors ``rockcraft_import::schema::ExtractedNote``."""

    pitch: int  # MIDI note number 0..=127
    start_us: int  # onset in microseconds from the start of the video
    dur_us: int  # duration in microseconds (>= 1 after conversion)
    hand: Hand = Hand.UNKNOWN
    velocity: Optional[int] = None  # None until M6-F audio-fusion fills it
    confidence: Optional[float] = None  # 0.0..=1.0, for the M6-E review step

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "pitch": int(self.pitch),
            "start_us": int(self.start_us),
            "dur_us": int(max(1, self.dur_us)),
            "hand": self.hand.value,
        }
        # Match serde's skip_serializing_if = Option::is_none.
        if self.velocity is not None:
            out["velocity"] = int(self.velocity)
        if self.confidence is not None:
            out["confidence"] = float(self.confidence)
        return out


@dataclass
class SourceMeta:
    """Provenance/diagnostics. Mirrors ``rockcraft_import::schema::SourceMeta``.

    ``audio_fusion`` and ``noise_filter`` are sidecar-only diagnostics recording
    what the M6-F audio pass and the visual noise-rejection stages did
    (``"applied: ..."`` / ``"skipped: ..."`` / ``"ghosts dropped N ..."``). The
    Rust deserializer does not ``deny_unknown_fields``, so it ignores these
    fields — they exist for logs and the M6-E review step, not for the Rust
    contract.
    """

    extractor_version: str = EXTRACTOR_VERSION
    title: Optional[str] = None
    fps: Optional[float] = None
    scroll_px_per_s: Optional[float] = None
    audio_fusion: Optional[str] = None
    noise_filter: Optional[str] = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"extractor_version": self.extractor_version}
        if self.title is not None:
            out["title"] = self.title
        if self.fps is not None:
            out["fps"] = float(self.fps)
        if self.scroll_px_per_s is not None:
            out["scroll_px_per_s"] = float(self.scroll_px_per_s)
        if self.audio_fusion is not None:
            out["audio_fusion"] = str(self.audio_fusion)
        if self.noise_filter is not None:
            out["noise_filter"] = str(self.noise_filter)
        return out


@dataclass
class ExtractedChart:
    """Top-level extractor output. Mirrors ``ExtractedChart``."""

    notes: list[ExtractedNote] = field(default_factory=list)
    source: SourceMeta = field(default_factory=SourceMeta)

    def to_dict(self) -> dict[str, Any]:
        return {
            "notes": [n.to_dict() for n in self.notes],
            "source": self.source.to_dict(),
        }

    def to_json(self, *, indent: Optional[int] = None) -> str:
        return json.dumps(self.to_dict(), indent=indent)

File: crates/import/src/schema.rs
use serde::{Deserialize, Serialize};

/// A chart extracted from a source by an importer sidecar.
///
/// This type is the wire format between a Python extractor and the Rust
/// pipeline — the M6-C video extractor and the M13-A score-file converter both
/// emit it. It must never be committed with real song data; synthetic fixtures
/// only (see M6-B gitignore guard).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ExtractedChart {
    pub notes: Vec<ExtractedNote>,
    pub source: SourceMeta,
    /// Notated musical context, when the source carries it. Score files do;
    /// video does not. Drives `meta.grid` / `meta.key` in the written bundle.
    ///
    /// `#[serde(default)]` keeps every chart written before M13-A — including
    /// the committed `tests/fixtures/synthetic_chart.json` — deserializable.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub notation: Option<NotationMeta>,
}

/// Tempo / metre / key as *notated* in the source score.
///
/// Only a source that states these explicitly should fill them in — nothing here
/// is inferred. A `None` field means "the source did not say", not "default".
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub struct NotationMeta {
    /// Tempo at the start of the piece, in BPM. The writer clamps it to
    /// [`Grid::MIN_BPM`](rockcraft_core::Grid::MIN_BPM)..=[`Grid::MAX_BPM`](rockcraft_core::Grid::MAX_BPM);
    /// out-of-range is not an error.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub bpm: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub time_sig: Option<rockcraft_core::TimeSig>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub key: Option<rockcraft_core::Key>,
}

/// One detected note in an extracted chart.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ExtractedNote {
    /// MIDI note number 0..=127.
    pub pitch: u8,
    /// Note start in microseconds from the beginning of the video.
    pub start_us: u64,
    /// Note duration in microseconds (at least 1 µs after conversion).
    pub dur_us: u64,
    pub hand: Hand,
    /// Explicit velocity 0..=127. `None` until M6-F audio-fusion fills it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub velocity: Option<u8>,
    /// Extractor confidence 0.0..=1.0. Used in the M6-E review step.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub confidence: Option<f32>,
}

/// Which hand played a note, as determined by the extractor.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum Hand {
    Left,
    Right,
    Unknown,
}

/// Provenance metadata emitted by the extractor sidecar.
///
/// Diagnostic only — never committed.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SourceMeta {
    /// Song title as inferred from the video, if available.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Video frame rate used during extraction.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fps: Option<f32>,
    /// Scroll speed in pixels per second used to convert positions to times.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub scroll_px_per_s: Option<f32>,
    /// Native source-frame height in pixels. With `hit_line_px` and
    /// `scroll_px_per_s`, the importer derives the edit overlay's grid
    /// calibration (`alignment.json`) so an imported chart registers on its
    /// source movie automatically. `None` when the extractor could not measure it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub frame_height_px: Option<u32>,
    /// Hit-line row in pixels from the top of the source frame (where falling
    /// notes are struck). Pairs with `frame_height_px` for the overlay calibration.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hit_line_px: Option<u32>,
    pub extractor_version: String,
}
